### ISSUE TITLE
Running multiple scans with ports specified causes validation to fail

### ISSUE THREAD

**Issue by noMoneyNoProblems** [jas-/node-libnmap#53]
Hey there,

First off, thanks for the great library. It really makes using nmap easier in node!

The guys on my team noticed a little bug that should be super simple to fix.

**Problem:**

Running two scans with the ports specified causes validation to fail with the following message:

> Trace: Error: Error: Error: Port(s) must match one of the following examples: 512 (single) | 0-65535 (range) | 10-30,80,443,3306-10000 (multiple)

**Cause:**

In lib/classes/validation.js the patterns.ports regexp produces alternating values when calling .test on the same input string.

The lines which trigger the error are:

`      if (!scope.patterns.ports.test(opts.ports))
        return cb(scope.messages.port);`

The regexp referred to: 

`      ports: /\d{1,6}(?:-\d+)?/g,`

This is because the /g flag is present on the end of the 'ports' regexp. See the following stack overflow page for an explanation: 

https://stackoverflow.com/questions/2630418/javascript-regex-returning-true-then-false-then-true-etc

**Reproduction code:**
```
const nmap = require('libnmap');
const opts1 = {
    flags: [ '-sV', '--version-light'],
    ports: "8080",
    range: [
        '127.0.0.1',
    ]
};

const opts2 = {
    flags: ['-sV', '--version-light'],
    ports: '5000',
    range: [
        '127.0.0.1',
    ]
};

 scan(opts1, "FIRST");

 scan(opts2, "SECOND");

function scan (opts, call) {
    nmap.scan(opts, function(err, report) {
        if (err) throw new Error(err);

        console.log("CALL IS: ", call)
        for (let item in report) {
            console.log(JSON.stringify(report[item], null, 2));
        }
    });
}

```
Call this file test_mult_libnmap.ts and run it with: `node test_mult_libnmap.ts`
The following output will be produced:

> Trace: Error: Error: Error: Port(s) must match one of the following examples: 512 (single) | 0-65535 (range) | 10-30,80,443,3306-10000 (multiple)
    at process.on (/home/exampleusername/exampleprojectfolder/exampleproject/node_modules/libnmap/lib/libnmap.js:116:17)
    at emitOne (events.js:116:13)
    at process.emit (events.js:211:7)
    at process._fatalException (bootstrap_node.js:374:26)
    at process.on (/home/exampleusername/exampleprojectfolder/exampleproject/node_modules/libnmap/lib/libnmap.js:116:11)
    at emitOne (events.js:116:13)
    at process.emit (events.js:211:7)
    at process._fatalException (bootstrap_node.js:374:26)

I'm not sure if that /g flag is necessary for the validation, or if it's extraneous. If the latter is true, it should be a single character fix! 

Thanks again, and hope to hear from you soon!

**Comment by jas-** [comment]
I am confused as to why you run two instances when you simply specify both ports and multiple nodes or networks to scan as demonstrated in the examples folder.

If there is a genuine reason for implementing your scan this way you may want to refactor.

**Comment by noMoneyNoProblems** [comment]
Hey jas-,

Thank you for the quick response.

If you'll allow me to explain our motivation for running multiple scans - we wish to monitor services running on individual ports, in a periodic fashion.

Ideally, as you say, we could scan all necessary ports at the same time. But it could be the case we need to scan different ports at different times, and therefore need separate scans.

As an example, imagine that we want to scan port 8080 of 192.158.133.1 every minute, and we also want to scan port 5000 and port 5001 of 192.158.133.2 every 2 minutes. Because these two scans need to run at different times, it is inappropriate to do both at the same time. But it's possible that two separate runs could overlap, hence the issue.

While experimenting briefly, we found multiple instances of libnmap could be run, as long as the patterns.ports regexp is tweaked.

More generally, there might be other cases where running two scan instances could be useful. For example, imagine a user of libnmap wanted to simultaneously run two scans with different flags set.

Excuse me for raising this issue if I've missed options that enable this behavior, or if such behavior is outside the scope of libnmap. 

And again, thank you for the time and effort you have spent on this library.

Cheers!

**Comment by jas-** [comment]
You could submit a pull request for this. It would have to pass ALL cases for the `ports` options; i.e. singular, ranges & mixed... (`npm test`)

**Comment by edgahan** [comment]
Just wanted to add this bug caught me out today as well. For what it's worth we don't run two instances at the same time, but a long running process that consumes from a queue and it was flip-flopping on the validations as above.

**Comment by jpstevens** [comment]
Alternative fix here - https://github.com/jas-/node-libnmap/pull/56

**Comment by jas-** [comment]
@jpstevens & @noMoneyNoProblems; thanks for bringing this to my attention.

I have addressed the two issues with a patch that should eliminate the problems with the `\g` when using the `str.test(regex)` function. 

The library was moved to a `str.match(regex)` method matching then re-assembling the `opts.ports` value as a string which is compatible with the [nmap](https:/nmap.org/book/man-port-specification.html) port specification syntax.

The latest published is now `v0.4.13`. I hope it fixes both of your problems.
